gateway/api/connections/connections.go

```diff
@@ -680,3 +693,379 @@ func GetDatabaseSchemas(c *gin.Context) {
+			if connStr := connEnvs["envvar:CONNECTION_STRING"]; connStr != "" {
+				dbName = getMongoDBFromConnectionString(connStr)
+			}
+		case pb.ConnectionTypeOracleDB:
+			dbName = getEnvValue(connEnvs, "envvar:SID")
+		}
+
+		if dbName == "" {
+			c.JSON(http.StatusBadRequest, gin.H{"message": "database name is required but not found in connection or query parameter"})
+			return
+		}
+	}
+
+	// Validate database name to prevent SQL injection
+	if err := validateDatabaseName(dbName); err != nil {
+		c.JSON(http.StatusUnprocessableEntity, gin.H{"message": err.Error()})
+		return
+	}
+
+	script := getTablesQuery(currentConnectionType, dbName)
+	if script == "" {
+		c.JSON(http.StatusBadRequest, gin.H{"message": "unsupported database type"})
+		return
+	}
+
+	userAgent := apiutils.NormalizeUserAgent(c.Request.Header.Values)
+	client, err := clientexec.New(&clientexec.Options{
+		OrgID:          ctx.GetOrgID(),
+		ConnectionName: conn.Name,
+		BearerToken:    getAccessToken(c),
+		UserAgent:      userAgent,
+		Verb:           pb.ClientVerbPlainExec,
+	})
+	if err != nil {
+		log.Error(err)
+		c.JSON(http.StatusInternalServerError, gin.H{"message": err.Error()})
+		return
+	}
+
+	respCh := make(chan *clientexec.Response)
+	go func() {
+		defer func() { close(respCh); client.Close() }()
+		select {
+		case respCh <- client.Run([]byte(script), nil):
+		default:
+		}
+	}()
+
+	timeoutCtx, cancelFn := context.WithTimeout(context.Background(), time.Second*30)
+	defer cancelFn()
+	select {
+	case outcome := <-respCh:
+		if outcome.ExitCode != 0 {
+			log.Errorf("failed issuing plain exec: %s, output=%v", outcome.String(), outcome.Output)
+			c.JSON(http.StatusInternalServerError, gin.H{"message": fmt.Sprintf("failed to list tables: %s", outcome.Output)})
+			return
+		}
+
+		response := TablesResponse{Schemas: []struct {
+			Name   string   `json:"name"`
+			Tables []string `json:"tables"`
+		}{}}
+
+		if currentConnectionType == pb.ConnectionTypeMongoDB {
+			// Parse MongoDB output
+			output := cleanMongoOutput(outcome.Output)
+			if output != "" {
+				var result []map[string]interface{}
+				if err := json.Unmarshal([]byte(output), &result); err != nil {
+					log.Errorf("failed parsing mongo response: %v", err)
+					c.JSON(http.StatusInternalServerError, gin.H{"message": fmt.Sprintf("failed to parse MongoDB response: %v", err)})
+					return
+				}
+
+				// Organize tables by schema
+				schemaMap := make(map[string][]string)
+				for _, row := range result {
+					schemaName := getString(row, "schema_name")
+					tableName := getString(row, "object_name")
+					schemaMap[schemaName] = append(schemaMap[schemaName], tableName)
+				}
+
+				// Convert map to response structure
+				for schemaName, tables := range schemaMap {
+					response.Schemas = append(response.Schemas, struct {
+						Name   string   `json:"name"`
+						Tables []string `json:"tables"`
+					}{
+						Name:   schemaName,
+						Tables: tables,
+					})
+				}
+			}
```

Comment:
Add this block to a function and create few unit test cases. I recommend unmarshalling the structure directly to the type if it's possible